<p style="text-align:center">
    <a href="https://skills.network/?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMDS0321ENSkillsNetwork26802033-2022-01-01" target="_blank">
    <img src="https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/assets/logos/SN_web_lightmode.png" width="200" alt="Skills Network Logo">
    </a>
</p>


# **Hands-on Lab: Interactive Visual Analytics with Folium**


Estimated time needed: **40** minutes


The launch success rate may depend on many factors such as payload mass, orbit type, and so on. It may also depend on the location and proximities of a launch site, i.e., the initial position of rocket trajectories. Finding an optimal location for building a launch site certainly involves many factors and hopefully we could discover some of the factors by analyzing the existing launch site locations.


In the previous exploratory data analysis labs, you have visualized the SpaceX launch dataset using `matplotlib` and `seaborn` and discovered some preliminary correlations between the launch site and success rates. In this lab, you will be performing more interactive visual analytics using `Folium`.


## Objectives


This lab contains the following tasks:

*   **TASK 1:** Mark all launch sites on a map
*   **TASK 2:** Mark the success/failed launches for each site on the map
*   **TASK 3:** Calculate the distances between a launch site to its proximities

After completed the above tasks, you should be able to find some geographical patterns about launch sites.


Let's first import required Python packages for this lab:


In [1]:
import piplite
await piplite.install(['folium'])
await piplite.install(['pandas'])

In [2]:
import folium
import pandas as pd

In [3]:
# Import folium MarkerCluster plugin
from folium.plugins import MarkerCluster
# Import folium MousePosition plugin
from folium.plugins import MousePosition
# Import folium DivIcon plugin
from folium.features import DivIcon

If you need to refresh your memory about folium, you may download and refer to this previous folium lab:


[Generating Maps with Python](https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-DV0101EN-SkillsNetwork/labs/v4/DV0101EN-Exercise-Generating-Maps-in-Python.ipynb)


In [7]:
import folium
from folium.features import DivIcon

nasa_coordinate = [29.559684888503615, -95.0830971930759]
site_map = folium.Map(location=nasa_coordinate, zoom_start=5)

for index, site in launch_sites_df.iterrows():
    coordinate = [site['Lat'], site['Long']]
    
    circle = folium.Circle(
        coordinate,
        radius=1000,
        color='#d35400',
        fill=True
    ).add_child(folium.Popup(site['Launch Site']))
    
    marker = folium.map.Marker(
        coordinate,
        icon=DivIcon(
            icon_size=(20, 20),
            icon_anchor=(0, 0),
            html='<div style="font-size: 12; color:#d35400;"><b>%s</b></div>' % site['Launch Site'],
        )
    )
    
    site_map.add_child(circle)
    site_map.add_child(marker)

site_map

First, let's try to add each site's location on a map using site's latitude and longitude coordinates


The following dataset with the name `spacex_launch_geo.csv` is an augmented dataset with latitude and longitude added for each site.


In [5]:
# Download and read the `spacex_launch_geo.csv`
from js import fetch
import io

URL = 'https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBM-DS0321EN-SkillsNetwork/datasets/spacex_launch_geo.csv'
resp = await fetch(URL)
spacex_csv_file = io.BytesIO((await resp.arrayBuffer()).to_py())
spacex_df=pd.read_csv(spacex_csv_file)

Now, you can take a look at what are the coordinates for each site.


In [6]:
# Select relevant sub-columns: `Launch Site`, `Lat(Latitude)`, `Long(Longitude)`, `class`
spacex_df = spacex_df[['Launch Site', 'Lat', 'Long', 'class']]
launch_sites_df = spacex_df.groupby(['Launch Site'], as_index=False).first()
launch_sites_df = launch_sites_df[['Launch Site', 'Lat', 'Long']]
launch_sites_df

,Launch Site,Lat,Long
0,CCAFS LC-40,28.562302,-80.577356
1,CCAFS SLC-40,28.563197,-80.576820
2,KSC LC-39A,28.573255,-80.646895
3,VAFB SLC-4E,34.632834,-120.610745


Above coordinates are just plain numbers that can not give you any intuitive insights about where are those launch sites. If you are very good at geography, you can interpret those numbers directly in your mind. If not, that's fine too. Let's visualize those locations by pinning them on a map.


We first need to create a folium `Map` object, with an initial center location to be NASA Johnson Space Center at Houston, Texas.


In [8]:
# Start location is NASA Johnson Space Center
nasa_coordinate = [29.559684888503615, -95.0830971930759]
site_map = folium.Map(location=nasa_coordinate, zoom_start=10)

We could use `folium.Circle` to add a highlighted circle area with a text label on a specific coordinate. For example,


In [9]:
# Create a blue circle at NASA Johnson Space Center's coordinate with a popup label showing its name
circle = folium.Circle(nasa_coordinate, radius=1000, color='#d35400', fill=True).add_child(folium.Popup('NASA Johnson Space Center'))
# Create a blue circle at NASA Johnson Space Center's coordinate with a icon showing its name
marker = folium.map.Marker(
    nasa_coordinate,
    # Create an icon as a text label
    icon=DivIcon(
        icon_size=(20,20),
        icon_anchor=(0,0),
        html='<div style="font-size: 12; color:#d35400;"><b>%s</b></div>' % 'NASA JSC',
        )
    )
site_map.add_child(circle)
site_map.add_child(marker)

and you should find a small yellow circle near the city of Houston and you can zoom-in to see a larger circle.


Now, let's add a circle for each launch site in data frame `launch_sites`


*Completed:* Create and add a `folium.Circle` and `folium.Marker` for each launch site.

The four launch sites are marked on `site_map` using their latitude/longitude coordinates, with the site name shown in the popup and as a map label.


An example of folium.Circle:


`folium.Circle(coordinate, radius=1000, color='#000000', fill=True).add_child(folium.Popup(...))`


An example of folium.Marker:


`folium.map.Marker(coordinate, icon=DivIcon(icon_size=(20,20),icon_anchor=(0,0), html='<div style="font-size: 12; color:#d35400;"><b>%s</b></div>' % 'label', ))`


In [10]:
# Initial the map
site_map = folium.Map(location=nasa_coordinate, zoom_start=5)

# Add a circle and marker for each launch site.
site_colors = ['lightblue', 'lightgreen', 'black', 'orange']
site_labels = {
    'CCAFS LC-40': 'CCAFS Launch Complex 40',
    'CCAFS SLC-40': 'CCAFS Space Launch Complex 40',
    'KSC LC-39A': 'KSC Launch Complex 39A',
    'VAFB SLC-4E': 'VAFB Space Launch Complex 4E'
}

for i, row in launch_sites_df.iterrows():
    coordinate = [row['Lat'], row['Long']]
    label = site_labels.get(row['Launch Site'], row['Launch Site'])
    circle = folium.Circle(
        coordinate,
        radius=1500,
        color=site_colors[i % len(site_colors)],
        fill=True
    ).add_child(folium.Popup(label))
    marker = folium.map.Marker(
        coordinate,
        icon=DivIcon(
            icon_size=(20, 20),
            icon_anchor=(0, 0),
            html='<div style="font-size: 12px; color:#d35400;"><b>%s</b></div>' % label
        )
    )
    site_map.add_child(circle)
    site_map.add_child(marker)

# Assign marker colors for Task 2
def assign_marker_color(launch_class):
    return 'green' if launch_class == 1 else 'red'

spacex_df['marker_color'] = spacex_df['class'].apply(assign_marker_color)

site_map


The generated map with marked launch sites should look similar to the following:


<center>
    <img src="https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBM-DS0321EN-SkillsNetwork/labs/module_3/images/launch_site_markers.png">
</center>


Now, you can explore the map by zoom-in/out the marked areas
, and try to answer the following questions:

*   Are all launch sites in proximity to the Equator line?
*   Are all launch sites in very close proximity to the coast?

### Answers

**1. Are all launch sites in proximity to the Equator line?**  
No. The launch sites are all in the Northern Hemisphere, but they are not especially close to the Equator. Three sites in Florida are around **28.6° N**, while VAFB SLC-4E in California is around **34.6° N**. This means they are roughly 28–35 degrees north of the Equator.

**2. Are all launch sites in very close proximity to the coast?**  
Yes. The mapped launch sites are located close to coastlines. The Florida launch sites are on/near the Atlantic coast, while VAFB SLC-4E is on the central California coast.

**Finding:** The launch sites are strongly associated with coastal areas, but their locations are not driven by being close to the Equator. Coastal placement is useful because rockets can launch over the ocean and reduce risk to populated land areas.


<class 'NameError'>: name 'marker_cluster' is not defined

Next, let's try to enhance the map by adding the launch outcomes for each site, and see which sites have high success rates.
Recall that data frame spacex_df has detailed launch records, and the `class` column indicates if this launch was successful or not


In [ ]:
spacex_df.tail(10)

Next, let's create markers for all launch records.
If a launch was successful `(class=1)`, then we use a green marker and if a launch was failed, we use a red marker `(class=0)`


Note that a launch only happens in one of the four launch sites, which means many launch records will have the exact same coordinate. Marker clusters can be a good way to simplify a map containing many markers having the same coordinate.


Let's first create a `MarkerCluster` object


In [12]:
marker_cluster = MarkerCluster()


*Completed:* Create a new `marker_color` column in `spacex_df` based on the `class` value.

- `class = 1` → `green` (successful launch)
- `class = 0` → `red` (failed launch)


In [13]:

# Apply a function to check the value of `class` column
# If class=1, marker_color value will be green
# If class=0, marker_color value will be red
# Task 2: Mark the success/failed launches for each site on the map
site_map.add_child(marker_cluster)

for index, record in spacex_df.iterrows():
    coordinate = [record['Lat'], record['Long']]
    marker = folium.Marker(
        coordinate,
        icon=folium.Icon(color='white', icon_color=record['marker_color'])
    )
    marker_cluster.add_child(marker)

site_map

*Completed:* For each launch result in `spacex_df`, add a `folium.Marker` to `marker_cluster`.


In [14]:
# Add marker_cluster to current site_map
site_map.add_child(marker_cluster)

# Add one marker for each launch record.
# Green indicates success (class=1); red indicates failure (class=0).
for index, record in spacex_df.iterrows():
    coordinate = [record['Lat'], record['Long']]
    marker = folium.Marker(
        coordinate,
        icon=folium.Icon(color='white', icon_color=record['marker_color'])
    )
    marker_cluster.add_child(marker)

site_map


Your updated map may look like the following screenshots:


<center>
    <img src="https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBM-DS0321EN-SkillsNetwork/labs/module_3/images/launch_site_marker_cluster.png">
</center>


<center>
    <img src="https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBM-DS0321EN-SkillsNetwork/labs/module_3/images/launch_site_marker_cluster_zoomed.png">
</center>


From the color-labeled markers in marker clusters, you should be able to easily identify which launch sites have relatively high success rates.

### Success-rate answer from `spacex_df`

| Launch site | Successful launches | Total launches | Success rate |
|---|---:|---:|---:|
| CCAFS LC-40 | 7 | 26 | 26.9% |
| CCAFS SLC-40 | 3 | 7 | 42.9% |
| KSC LC-39A | 10 | 13 | **76.9%** |
| VAFB SLC-4E | 4 | 10 | 40.0% |

**Finding:** `KSC LC-39A` has the highest observed success rate in this dataset (about **76.9%**), while `CCAFS LC-40` has the lowest (about **26.9%**).

In [15]:
# TASK 3: Calculate the distances between a launch site to its proximities


<class 'NameError'>: name 'calculate_distance' is not defined

Next, we need to explore and analyze the proximities of launch sites.


Let's first add a `MousePosition` on the map to get coordinate for a mouse over a point on the map. As such, while you are exploring the map, you can easily find the coordinates of any points of interests (such as railway)


In [ ]:
# Add Mouse Position to get the coordinate (Lat, Long) for a mouse over on the map
formatter = "function(num) {return L.Util.formatNum(num, 5);};"
mouse_position = MousePosition(
    position='topright',
    separator=' Long: ',
    empty_string='NaN',
    lng_first=False,
    num_digits=20,
    prefix='Lat:',
    lat_formatter=formatter,
    lng_formatter=formatter,
)

site_map.add_child(mouse_position)
site_map

Now zoom in to a launch site and explore its proximity to see if you can easily find any railway, highway, coastline, etc. Move your mouse to these points and mark down their coordinates (shown on the top-left) in order to the distance to the launch site.


Now zoom in to a launch site and explore its proximity to see if you can easily find any railway, highway, coastline, etc. Move your mouse to these points and mark down their coordinates (shown on the top-left) in order to the distance to the launch site.


In [ ]:
from math import sin, cos, sqrt, atan2, radians

def calculate_distance(lat1, lon1, lat2, lon2):
    # approximate radius of earth in km
    R = 6373.0

    lat1 = radians(lat1)
    lon1 = radians(lon1)
    lat2 = radians(lat2)
    lon2 = radians(lon2)

    dlon = lon2 - lon1
    dlat = lat2 - lat1

    a = sin(dlat / 2)**2 + cos(lat1) * cos(lat2) * sin(dlon / 2)**2
    c = 2 * atan2(sqrt(a), sqrt(1 - a))

    distance = R * c
    return distance

*Completed:* For the VAFB SLC-4E example, use the closest coastline point identified with `MousePosition` and calculate the distance.

Closest coastline point used by this notebook example:
- Latitude: **34.63462**
- Longitude: **-120.62439**
- Distance from VAFB SLC-4E: **1.26 km**


In [ ]:
# Coordinates of the selected closest coastline point for VAFB SLC-4E
coastline_lat = 34.63462
coastline_lon = -120.62439

launch_site_lat = launch_sites_df.loc[launch_sites_df['Launch Site'] == 'VAFB SLC-4E', 'Lat'].iloc[0]
launch_site_lon = launch_sites_df.loc[launch_sites_df['Launch Site'] == 'VAFB SLC-4E', 'Long'].iloc[0]

distance_coastline = calculate_distance(
    launch_site_lat, launch_site_lon,
    coastline_lat, coastline_lon
)

print(f"Distance to coastline: {distance_coastline:.2f} km")


In [ ]:
# Add a marker at the selected coastline point and display the distance.
coastline_coordinate = [coastline_lat, coastline_lon]

distance_marker = folium.Marker(
    coastline_coordinate,
    icon=DivIcon(
        icon_size=(20, 20),
        icon_anchor=(0, 0),
        html='<div style="font-size: 12px; color:#d35400;"><b>%s</b></div>'
              % "{:10.2f} KM".format(distance_coastline),
    )
)


*Completed:* Draw a `PolyLine` between VAFB SLC-4E and the selected coastline point.


In [ ]:
# Draw a line from the VAFB SLC-4E launch site to the selected coastline point.
launch_site_coordinate = [launch_site_lat, launch_site_lon]
lines = folium.PolyLine(
    locations=[coastline_coordinate, launch_site_coordinate],
    weight=2
)
site_map.add_child(lines)
site_map.add_child(distance_marker)


Your updated map with distance line should look like the following screenshot:


<center>
    <img src="https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBM-DS0321EN-SkillsNetwork/labs/module_3/images/launch_site_marker_distance.png">
</center>


*Completed:* The example below draws distance lines from VAFB SLC-4E to nearby railway, highway, and city points obtained with `MousePosition`.


A railway map symbol may look like this:


<center>
    <img src="https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBM-DS0321EN-SkillsNetwork/labs/module_3/images/railway.png">
</center>


A highway map symbol may look like this:


<center>
    <img src="https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBM-DS0321EN-SkillsNetwork/labs/module_3/images/highway.png">
</center>


A city map symbol may look like this:


<center>
    <img src="https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBM-DS0321EN-SkillsNetwork/labs/module_3/images/city.png">
</center>


In [ ]:
# Create markers with distances to the closest railway, highway, and city.
# Draw a line from each proximity point to VAFB SLC-4E.
railway_coordinates = coastline_coordinate
highway_coordinates = (34.68577, -120.59585)
city_coordinates = (34.63886, -120.45788)

distance_railway = distance_coastline
distance_highway = calculate_distance(
    launch_site_lat, launch_site_lon,
    highway_coordinates[0], highway_coordinates[1]
)
distance_city = calculate_distance(
    launch_site_lat, launch_site_lon,
    city_coordinates[0], city_coordinates[1]
)

dist_railway_marker = folium.Marker(
    railway_coordinates,
    icon=DivIcon(
        icon_size=(20,20),
        icon_anchor=(0,0),
        html='<div style="font-size: 12px; color:#d35400;"><b>%s</b></div>'
              % "{:10.2f} KM".format(distance_railway),
    )
)

dist_highway_marker = folium.Marker(
    highway_coordinates,
    icon=DivIcon(
        icon_size=(20,20),
        icon_anchor=(0,0),
        html='<div style="font-size: 12px; color:#d35400;"><b>%s</b></div>'
              % "{:10.2f} KM".format(distance_highway),
    )
)

dist_city_marker = folium.Marker(
    city_coordinates,
    icon=DivIcon(
        icon_size=(20,20),
        icon_anchor=(0,0),
        html='<div style="font-size: 12px; color:#d35400;"><b>%s</b></div>'
              % "{:10.2f} KM".format(distance_city),
    )
)


In [ ]:
# Draw the distance lines.
railway_lines = folium.PolyLine(
    locations=[railway_coordinates, launch_site_coordinate], weight=2
)
highway_lines = folium.PolyLine(
    locations=[highway_coordinates, launch_site_coordinate], weight=2
)
city_lines = folium.PolyLine(
    locations=[city_coordinates, launch_site_coordinate], weight=2
)

site_map.add_child(railway_lines)
site_map.add_child(highway_lines)
site_map.add_child(city_lines)


In [ ]:
# Add the distance labels to the map.
site_map.add_child(dist_railway_marker)
site_map.add_child(dist_highway_marker)
site_map.add_child(dist_city_marker)

site_map


After you plot distance lines to the proximities, you can answer the following questions easily:

*   Are launch sites in close proximity to railways?
*   Are launch sites in close proximity to highways?
*   Are launch sites in close proximity to coastline?
*   Do launch sites keep certain distance away from cities?

### Answers for the VAFB SLC-4E example

- **Railway:** Yes. The selected railway point is about **1.26 km** from the launch site, so it is in close proximity.
- **Highway:** The selected highway point is about **6.04 km** away. This is reasonably close, but farther than the railway and coastline points.
- **Coastline:** Yes. The selected coastline point is about **1.26 km** away, showing that VAFB SLC-4E is very close to the coast.
- **Cities:** The selected city point is about **14.01 km** away. This is noticeably farther than the railway, highway, and coastline.

**Finding:** The VAFB example indicates a pattern of placing a launch site close to the coastline and transportation infrastructure while maintaining a larger separation from nearby cities. These distances are for the specific proximity points selected in the lab, so they should not be treated as exact measurements for every launch site.


### Completed analysis note
The answers above are based on the launch-site coordinates and launch outcome data used by this lab. The proximity measurements use the VAFB SLC-4E example points selected in the lab.


# Next Steps:

Now you have discovered many interesting insights related to the launch sites' location using folium, in a very interactive way. Next, you will need to build a dashboard using Ploty Dash on detailed launch records.


## Authors


[Pratiksha Verma](https://www.linkedin.com/in/pratiksha-verma-6487561b1/)


<!--## Change Log--!>


<!--| Date (YYYY-MM-DD) | Version | Changed By      | Change Description      |
| ----------------- | ------- | -------------   | ----------------------- |
| 2022-11-09        | 1.0     | Pratiksha Verma | Converted initial version to Jupyterlite|--!>


### <h3 align="center"> IBM Corporation 2022. All rights reserved. <h3/>
